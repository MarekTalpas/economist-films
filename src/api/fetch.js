// @flow
function constructFetch(type: string, id: ?number) {
  const urlString = id ? `${type}/${id}` : type;
  return fetch(`
    https://economist.twivel.io/api/v1/${urlString}/json`, {
      method: 'GET',
    })
  .then(response => response.json())
  .catch(err => err);
}

export function getRecommendedEpisodes(id: number) {
  return fetch(`http://ec2-52-59-206-213.eu-central-1.compute.amazonaws.com:8080/user/${id}`,
    {
      method: 'GET',
    })
  .then(response => response.json())
  .catch(err => err);
}

export function getRoot() {
  return constructFetch('root');
}

export function getEpisodeByID(id: number) {
  return constructFetch('episode', id);
}

export function getSeriesByID(id: number) {
  return constructFetch('series', id);
}
