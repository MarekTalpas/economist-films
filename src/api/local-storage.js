// @flow
import _ from 'lodash';

export type VideoProgressType = {
  episodeId: number,
  progressTime: number,
};

export const HISTORY_LIST = 'history';

export function saveVideoProgress(id: number, progress: number) {
  if (typeof (localStorage) !== 'undefined') {
    const currentHistory: ?string = localStorage.getItem(HISTORY_LIST);
    const parsedCurrentHistory = JSON.parse(currentHistory || 'null') || [];
    const indexOfVideo = parsedCurrentHistory.findIndex(element => element.episodeId === id);
    if (indexOfVideo === -1) {
      parsedCurrentHistory.push({
        episodeId: id,
        progressTime: progress,
      });
    } else {
      parsedCurrentHistory[indexOfVideo].progressTime = progress;
    }
    localStorage.setItem(HISTORY_LIST, JSON.stringify(parsedCurrentHistory));
  }
}

let counter = 0;

export function getProgressTimeById(id: number) {
  const data: ?string = localStorage.getItem(HISTORY_LIST);
  const parsedData = JSON.parse(data || 'null');
  if (parsedData) {
    const requiredEpisode = parsedData.find(episode => episode.episodeId === id);
    return requiredEpisode ? requiredEpisode.progressTime : 0;
  }
  return 0;
}

export function getLastWatchedEpisodeID(episodeIDs: Array<number>) {
  const endBoundary = 95;
  const data: ?string = localStorage.getItem(HISTORY_LIST);
  const parsedData = JSON.parse(data || 'null');
  if (parsedData) {
    const episodesCount = episodeIDs.length;
    const commonEpisodes =
      parsedData.filter(episode => episodeIDs.find(ids => (ids === episode.episodeId)));
    const mappedParsedData = commonEpisodes.filter(video => video.progressTime >= 95);
    if (mappedParsedData.length === episodesCount) {
      const differentEpisodes = _.differenceBy(parsedData, commonEpisodes, 'episodeId');
      localStorage.setItem(HISTORY_LIST, JSON.stringify(differentEpisodes));
    }
  }
  const firstEpisode = episodeIDs[0];
  const lastFinishedId = episodeIDs[counter];
  const lastEpisode =
    parsedData.find(histEntry => histEntry.episodeId === lastFinishedId);
  if (lastEpisode) {
    if (lastEpisode.progressTime < endBoundary) {
      return lastFinishedId || firstEpisode;
    }
    counter += 1;
    return episodeIDs[counter] || firstEpisode;
  }
  return lastFinishedId || firstEpisode;
}
