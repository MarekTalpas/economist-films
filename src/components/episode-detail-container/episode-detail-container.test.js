// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { getSeriesByID, getEpisodeByID } from '../../api/fetch';
import EpisodeDetailContainer from './episode-detail-container';

test('EpisodeDetailContainer renders correctly', () => {
  const tree : string = renderer.create(<EpisodeDetailContainer params />).toJSON();
  expect(tree).toMatchSnapshot();
});
