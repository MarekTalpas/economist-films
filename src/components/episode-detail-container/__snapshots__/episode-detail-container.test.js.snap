exports[`test EpisodeDetailContainer renders correctly 1`] = `
<div
  className="episode-detail">
  <div
    className="episode-detail__details">
    <div
      className="episode-watch">
      <div
        className="episode-watch__video"
        style={
          Object {
            "background": "#000 url(\"undefined\") center/cover no-repeat",
          }
        } />
      <a
        className="episode-watch__watchnow-button"
        onClick={[Function]}
        style={Object {}}>
        Watch now
      </a>
    </div>
    <div
      className="episode-description">
      <h3
        className="episode-description__date" />
      <h1
        className="episode-description__title" />
      <p
        className="episode-description__text" />
    </div>
  </div>
  <div
    className="episode-detail__recommended">
    <h1
      className="episode-detail__recommended-title">
      Recommended
    </h1>
    <div
      className="episode-detail__recommended-card-wrapper" />
  </div>
</div>
`;
